File: js_tools/grep.js
return function(args) {
  if (!args || typeof args.pattern !== "string" || args.pattern === "") {
    throw new Error("INVALID_ARGUMENT: Missing mandatory field: pattern");
  }

  const pattern = args.pattern;
  const searchPath = args.path || ".";

  function parseOptionalInteger(value, fallback) {
    if (value === undefined || value === null) return fallback;
    if (typeof value === "number" && Number.isInteger(value)) return value;
    if (typeof value === "string" && /^-?\d+$/.test(value.trim())) return parseInt(value.trim(), 10);
    return fallback;
  }

  function splitIgnoreInput(value) {
    if (value === undefined || value === null) return [];
    if (Array.isArray(value)) return value.filter(function(v) { return typeof v === "string" && v.trim() !== ""; });
    if (typeof value === "string") {
      return value.split(",").map(function(v) { return v.trim(); }).filter(function(v) { return v !== ""; });
    }
    return [];
  }

  function uniquePush(list, value) {
    if (value && list.indexOf(value) === -1) list.push(value);
  }

  function resolveRepoRoot() {
    const res = __os_run("git rev-parse --show-toplevel");
    if (res.exit_code !== 0) return null;
    const out = (res.stdout || "").trim();
    return out === "" ? null : out;
  }

  function parseRootGitignore(repoRoot) {
    const result = { dirs: [], files: [], skipped: 0 };
    if (!repoRoot) return result;

    const filePath = repoRoot + "/.gitignore";
    const check = __os_run("test -f " + shell_escape(filePath));
    if (check.exit_code !== 0) return result;

    const cat = __os_run("cat " + shell_escape(filePath));
    if (cat.exit_code !== 0) return result;

    const lines = (cat.stdout || "").split("\n");
    for (let i = 0; i < lines.length; i++) {
      const raw = lines[i];
      const line = raw.trim();
      if (line === "" || line[0] === "#") continue;

      if (line[0] === "!" || line.indexOf("**") !== -1) {
        result.skipped += 1;
        continue;
      }

      if (line[line.length - 1] === "/") {
        const dir = line.slice(0, -1).trim();
        if (dir === "" || dir.indexOf("/") !== -1) {
          result.skipped += 1;
          continue;
        }
        uniquePush(result.dirs, dir);
        continue;
      }

      if (line.indexOf("/") !== -1) {
        result.skipped += 1;
        continue;
      }

      uniquePush(result.files, line);
    }

    return result;
  }

  const limit = Math.max(1, parseOptionalInteger(args.limit, 200));
  const context = Math.max(0, parseOptionalInteger(args.context, 0));
  const contextArg = context > 0 ? " -C " + context : "";

  const defaultIgnores = [".git", "node_modules", "bazel-*", "dist", "build", ".cache", "target"];
  const userIgnores = splitIgnoreInput(args.ignore);
  const includeIgnored = args.include_ignored === true;

  const repoRoot = resolveRepoRoot();
  const gitignore = parseRootGitignore(repoRoot);

  const excludeDirs = [];
  const excludeFiles = [];

  if (!includeIgnored) {
    for (let i = 0; i < defaultIgnores.length; i++) uniquePush(excludeDirs, defaultIgnores[i]);
  }
  for (let i = 0; i < userIgnores.length; i++) uniquePush(excludeDirs, userIgnores[i]);
  if (!includeIgnored) {
    for (let i = 0; i < gitignore.dirs.length; i++) uniquePush(excludeDirs, gitignore.dirs[i]);
  }
  if (!includeIgnored) {
    for (let i = 0; i < gitignore.files.length; i++) uniquePush(excludeFiles, gitignore.files[i]);
  }

  const excludeArgs = [];
  for (let i = 0; i < excludeDirs.length; i++) excludeArgs.push("--exclude-dir=" + shell_escape(excludeDirs[i]));
  for (let i = 0; i < excludeFiles.length; i++) excludeArgs.push("--exclude=" + shell_escape(excludeFiles[i]));

  const modeFlag = args.fixed_strings === true ? "-F" : "-E";
  const cmd = "grep -rn " + modeFlag + " -I --color=never" +
    contextArg +
    (excludeArgs.length ? " " + excludeArgs.join(" ") : "") +
    " -e " + shell_escape(pattern) +
    " " + shell_escape(searchPath);

  const res = __os_run(cmd);
  if (res.exit_code !== 0 && res.exit_code !== 1) {
    throw new Error("INTERNAL: Command failed with status " + res.exit_code + "\nOutput:\n" + (res.stdout || "") + (res.stderr || ""));
  }

  const stdout = res.stdout || "";
  const lines = stdout === "" ? [] : stdout.split("\n");
  let output = stdout;

  if (lines.length > limit) {
    output = lines.slice(0, limit).join("\n") +
      "\n[TRUNCATED: Use a more specific pattern or path to narrow results]";
  }

  if (gitignore.skipped > 0) {
    output = output + (output === "" ? "" : "\n") +
      "[NOTE: Skipped " + gitignore.skipped + " unsupported root .gitignore rule(s)]";
  }

  return output;
};

File: core/tools/grep.cpp
#include "core/tool_executor.h"

#include <sstream>
#include <vector>

#include "absl/status/status.h"
#include "absl/strings/ascii.h"
#include "absl/strings/match.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_join.h"

#include "core/shell_util.h"
#include "core/status_macros.h"
#include "core/tools/common.h"
#include "core/json_utils.h"

namespace slop {
absl::StatusOr<std::string> ToolExecutor::HandleGrep(const nlohmann::json& args) {
  auto pattern = json_get<std::string>(args, "pattern");
  if (!pattern || pattern->empty()) {
    return absl::InvalidArgumentError("Missing mandatory field: pattern");
  }

  const std::string search_path = json_get_or<std::string>(args, "path", ".");

  ASSIGN_OR_RETURN(auto context_opt, ParseOptionalInteger(args, "context"));
  ASSIGN_OR_RETURN(auto limit_opt, ParseOptionalInteger(args, "limit"));
  const int context = std::max(0, context_opt.value_or(0));
  const int limit = std::max(1, limit_opt.value_or(200));

  std::vector<std::string> ignore_dirs;
  std::vector<std::string> ignore_files;
  const bool include_ignored = json_get_or<bool>(args, "include_ignored", false);
  if (!include_ignored) {
    ignore_dirs = {".git", "node_modules", "bazel-*", "dist", "build", ".cache", "target"};
  }

  if (const auto* ignore = json_at(args, "ignore")) {
    if (ignore->is_string() && !ignore->get<std::string>().empty()) {
      ignore_dirs.push_back(ignore->get<std::string>());
    } else if (ignore->is_array()) {
      for (const auto& v : *ignore) {
        if (v.is_string() && !v.get<std::string>().empty()) {
          ignore_dirs.push_back(v.get<std::string>());
        }
      }
    }
  }

  // Minimal root .gitignore support parity: plain dir/filename entries only.
  int skipped_gitignore_rules = 0;
  if (!include_ignored) {
    auto root_or = RunCommand("git rev-parse --show-toplevel 2>/dev/null");
    if (root_or.ok() && root_or->exit_code == 0) {
      std::string repo_root = TrimNewlines(root_or->stdout_out);
      if (!repo_root.empty()) {
        auto cat_or = RunCommand(absl::StrCat("cat ", EscapeShellArg(absl::StrCat(repo_root, "/.gitignore")), " 2>/dev/null"));
        if (cat_or.ok() && cat_or->exit_code == 0) {
          std::stringstream gs(cat_or->stdout_out);
          std::string line;
          while (std::getline(gs, line)) {
            std::string t(absl::StripAsciiWhitespace(line));
            if (t.empty() || t[0] == '#') continue;
            if (t[0] == '!' || absl::StrContains(t, "**")) {
              skipped_gitignore_rules++;
              continue;
            }
            if (!t.empty() && t.back() == '/') {
              t.pop_back();
              t = absl::StripAsciiWhitespace(t);
              if (t.empty() || absl::StrContains(t, "/")) {
                skipped_gitignore_rules++;
              } else {
                ignore_dirs.push_back(t);
              }
              continue;
            }
            if (absl::StrContains(t, "/")) {
              skipped_gitignore_rules++;
            } else {
              ignore_files.push_back(t);
            }
          }
        }
      }
    }
  }

  std::vector<std::string> exclude_args;
  exclude_args.reserve(ignore_dirs.size() + ignore_files.size());
  for (const auto& d : ignore_dirs) exclude_args.push_back(absl::StrCat("--exclude-dir=", EscapeShellArg(d)));
  for (const auto& f : ignore_files) exclude_args.push_back(absl::StrCat("--exclude=", EscapeShellArg(f)));

  const bool fixed_strings = json_get_or<bool>(args, "fixed_strings", false);
  const std::string mode_flag = fixed_strings ? "-F" : "-E";
  const std::string context_arg = context > 0 ? absl::StrCat(" -C ", context) : "";
  const std::string cmd = absl::StrCat("grep -rn ", mode_flag, " -I --color=never", context_arg,
                                       (exclude_args.empty() ? "" : absl::StrCat(" ", absl::StrJoin(exclude_args, " "))),
                                       " -e ", EscapeShellArg(*pattern), " ", EscapeShellArg(search_path));

  ASSIGN_OR_RETURN(auto res, RunCommand(cmd));
  if (res.exit_code != 0 && res.exit_code != 1) {
    return absl::InternalError(
        absl::StrCat("INTERNAL: Command failed with status ", res.exit_code, "\nOutput:\n", res.stdout_out, res.stderr_out));
  }

  std::string output = res.stdout_out;
  std::vector<std::string> lines;
  if (!output.empty()) {
    std::stringstream os(output);
    std::string l;
    while (std::getline(os, l)) lines.push_back(l);
  }
  if (static_cast<int>(lines.size()) > limit) {
    output = absl::StrCat(absl::StrJoin(std::vector<std::string>(lines.begin(), lines.begin() + limit), "\n"),
                          "\n[TRUNCATED: Use a more specific pattern or path to narrow results]");
  }
  if (skipped_gitignore_rules > 0) {
    if (!output.empty() && output.back() == '\n') output.pop_back();
    output = absl::StrCat(output, (output.empty() ? "" : "\n"), "[NOTE: Skipped ", skipped_gitignore_rules,
                          " unsupported root .gitignore rule(s)]");
  }
  return output;
}

}  // namespace slop
